EIPS/eip-5007.md

```diff
@@ -0,0 +1,313 @@
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+```solidity
+interface IERC5007  {
+
+    /// @notice Emitted when the `startTime` or `endTime` of a NFT is changed
+    /// @param tokenId  The tokenId of the NFT
+    /// @param startTime  The new start time of the NFT
+    /// @param endTime  The new end time of the NFT
+    event TimeUpdate(uint256 tokenId, uint64 startTime, uint64 endTime);
+
+    /// @notice Get the start time of the NFT
+    /// @dev Throws if `tokenId` is not valid NFT
+    /// @param tokenId  The tokenId of the NFT
+    /// @return The start time of the NFT
+    function startTime(uint256 tokenId) external view returns (uint64);
+    
+    /// @notice Get the end time of the NFT
+    /// @dev Throws if `tokenId` is not valid NFT
+    /// @param tokenId  The tokenId of the NFT
+    /// @return The end time of the NFT
+    function endTime(uint256 tokenId) external view returns (uint64);
+
+    /// @notice Get the token id which this NFT mint from
+    /// @dev Throws if `tokenId` is not valid NFT
+    /// @param tokenId  The tokenId of the NFT
+    /// @return The token id which this NFT mint from
+    function originalTokenId(uint256 tokenId) external view returns (uint256);
+
+    /// @notice Check the NFT is valid now 
+    /// @dev Throws if `tokenId` is not valid NFT
+    /// @param tokenId  The tokenId of the NFT
+    /// @return The the NFT is valid now
+    /// if(startTime <= now <= endTime) {return true;} else {return false;}
+    function isValidNow(uint256 tokenId) external view returns (bool);
+
+    /// @notice Mint a new token from an old token
+    /// @dev Throws if `tokenId` is not valid token
+    /// @param originalTokenId_  The token id which the new token mint from
+    /// @param newTokenStartTime  The start time of the new token
+    /// @param newTokenOwner  The owner of the new token
+    /// @return newTokenId The the token id of the new token
+    function split(uint256 originalTokenId_, uint64 newTokenStartTime, address newTokenOwner) external returns(uint256);
+
+    /// @notice Merge two time NFTs into one time NFT
+    /// @dev Throws if `firstTokenId` or `secondTokenId` is not valid token
+    /// @param firstTokenId   The id of the first token
+    /// @param secondTokenId  The id of the second token
+    /// @param newTokenOwner  The owner of the new token
+    /// @return newTokenId The id of the new token
+    function merge(uint256 firstTokenId, uint256 secondTokenId, address newTokenOwner) external returns(uint256);
+}
+```
+
+## Rationale
+
+### Time Data Type
+
+The max value of `uint64` is 18446744073709551615,  timestamp 18446744073709551615 is about year 584942419325. `uint64` is sufficient and saves gas compared to `uint256`.
```

Comment:
Does this actually save gas? I thought all integer types encoded to 32 bytes: https://docs.soliditylang.org/en/develop/abi-spec.html#examples

Reply:
> Does this actually save gas? I thought all integer types encoded to 32 bytes: 

``` struct TimeNftInfo {  uint256 originalTokenId;  uint64 startTime;  uint64 endTime; }```  encoded to 64 bytes.

``` struct TimeNftInfo {  uint256 originalTokenId;  uint256 startTime;  uint256 endTime; }```  encoded to 96 bytes.

Reply:
Another reason，`uint256` is too big for some software  such as MySQL、Elastic Search.  `uint64` is natively supported on many software and programming languages.

Reply:
> Another reason，uint256 is too big for some software such as MySQL, Elastic Search. uint64 is natively supported on many software and programming languages.

That's still completely arbitrary.

I just ran some tests on a project that already makes heavy use of time-based NFTs to confirm whether it was even worth it to switch to uint64 for storage. Our gas costs actually ended going up...

I highly recommend switching to uint256 completely.

Reply:
> I just ran some tests on a project that already makes heavy use of time-based NFTs to confirm whether it was even worth it to switch to uint64 for storage. Our gas costs actually ended going up...

Could you please send the code link of this  time-based NFTs?

Reply:
It's private at the moment, but the reasoning for can be elaborated on. There is still some minimal amount of arithmetic and comparisons against other uint256's, and the uint64 => uint256 transitions adds gas everytime. For example, a common use-case is to add or compare against *now* (`block.timestamp`) which is natively `uint256`.